sdk/search/azure-search-documents/azure/search/documents/_internal/_search_index_document_batching_client.py

```diff
@@ -0,0 +1,280 @@
+        """
+        actions = self._index_documents_batch.add_delete_actions(documents)
+        if self._persistence:
+            self._persistence.add_queued_actions(actions)
+        self._flush_if_needed()
+
+    def merge_documents_actions(self, documents):
+        # type: (List[dict]) -> None
+        """Queue merge documents actions
+
+        :param documents: A list of documents to merge.
+        :type documents: List[dict]
+        """
+        actions = self._index_documents_batch.add_merge_actions(documents)
+        if self._persistence:
+            self._persistence.add_queued_actions(actions)
+        self._flush_if_needed()
+
+    def merge_or_upload_documents_actions(self, documents):
+        # type: (List[dict]) -> None
+        """Queue merge documents or upload documents actions
+
+        :param documents: A list of documents to merge or upload.
+        :type documents: List[dict]
+        """
+        actions = self._index_documents_batch.add_merge_or_upload_actions(documents)
+        if self._persistence:
+            self._persistence.add_queued_actions(actions)
+        self._flush_if_needed()
+
+    @distributed_trace
+    def _index_documents_actions(self, actions, **kwargs):
+        # type: (List[IndexAction], **Any) -> List[IndexingResult]
+        error_map = {413: RequestEntityTooLargeError}
+
+        kwargs["headers"] = self._merge_client_headers(kwargs.get("headers"))
+        try:
+            index_documents = IndexBatch(actions=actions)
+            batch_response = self._client.documents.index(batch=index_documents, error_map=error_map, **kwargs)
+            return cast(List[IndexingResult], batch_response.results)
+        except RequestEntityTooLargeError:
+            if len(actions) == 1:
+                raise
+            pos = round(len(actions) / 2)
+            batch_response_first_half = self._index_documents_actions(
+                actions=actions[:pos],
+                error_map=error_map,
+                **kwargs
+            )
+            if batch_response_first_half:
+                result_first_half = cast(List[IndexingResult], batch_response_first_half.results)
+            else:
+                result_first_half = []
+            batch_response_second_half = self._index_documents_actions(
+                actions=actions[pos:],
+                error_map=error_map,
+                **kwargs
+            )
+            if batch_response_second_half:
+                result_second_half = cast(List[IndexingResult], batch_response_second_half.results)
+            else:
+                result_second_half = []
+            return result_first_half.extend(result_second_half)
+
+    def __enter__(self):
+        # type: () -> SearchClient
+        self._client.__enter__()  # pylint:disable=no-member
+        return self
+
+    def __exit__(self, *args):
+        # type: (*Any) -> None
+        self._client.__exit__(*args)  # pylint:disable=no-member
```

Comment:
so shall we clean-up and/or close and/or flush? what's the close story?

Reply:
The concern is the flush method may raise errors and in some cases, user may want to catch the error and take actions (e.g. save the state).